### PR TITLE
v2: Context-aware WaitFor

### LINKED ISSUE #2956: v2: Context-Aware WaitFor
With the adoption of contexts WaitFor should also be context-aware.

### PR DESCRIPTION
**Pull request by vooon**
Fixes #2956

### CODE REVIEW

**Comment by pierreprinetti** [comment]
Good catch, thanks.

I would remove the argument in seconds and let the caller pass a context with a deadline. WDYT?

**Comment by vooon** [comment]
@pierreprinetti indeed in many cases context with the deadline is enough, but i found it handy option to pass a timeout.

Context of the predicate anyway should be canceled once we exit WaitFor, so it's just handy way of use either WithCancel or WithTimeout if timeout > 0.

I'm unsure about other WaitFor* funcs, unfortunately my regex-fu is not that great...

**Comment by pierreprinetti** [comment]
> i found it handy option to pass a timeout

The `context` library [provides a convenience function for setting a timeout](https://pkg.go.dev/context#WithTimeout). Note that if you set a deadline or timeout on a context that already has one, the shortest delay correctly wins.

**Comment by pierreprinetti** [comment]
My stance here is: this is not Gophercloud's job. There are plenty of libraries out there that do Wait in fanciful ways. I understand that we might continue providing a function that covers the basic case, but we should keep complexity at a minimum to avoid bugs. Accepting a context and letting the user set it appropriately seems to me like a reasonable compromise.

**Comment by vooon** [comment]
@pierreprinetti agree, that surface should be minimal but enough. I also agree, that usually you just pass deadline ctx and do not expect fancy logic inside.

Optional timeout just makes tests a bit easier to write, but i often have limited context in test suite anyway.

I'll drop timeouts.

**Comment by vooon** [comment]
@pierreprinetti it's pretty much done, but unsure why is the CI fail.

**Comment by pierreprinetti** [comment]
There's a race. Your select statement is a bit funky, in that it could trigger an unbound number of predicate calls. Your implementation makes sure that a function is called each second even though the previous one hasn't returned. Which is interesting, but I don't think it's useful in our case.

I believe that you tied yourself in knots by building on the old function... but if we start relying just on context, we might get away with something very simple:

```Go
// WaitFor polls a predicate function, once per second, up to a context cancellation.
// This is useful to wait for a resource to transition to a certain state.
// Resource packages will wrap this in a more convenient function that's
// specific to a certain resource, but it can also be useful on its own.
func WaitFor(ctx context.Context, predicate func(context.Context) (bool, error)) error {
	if done, err := predicate(ctx); done || err != nil {
		return err
	}

	ticker := time.NewTicker(1 * time.Second)
	defer ticker.Stop()

	for {
		select {
		case <-ticker.C:
			if done, err := predicate(ctx); done || err != nil {
				return err
			}
		case <-ctx.Done():
			return ctx.Err()
		}
	}
}
```

**Comment by vooon** [comment]
@pierreprinetti yes, that looks a lot better.

**Comment by vooon** [comment]
@pierreprinetti okay, it's finally passes most of the tests. Still unclear why functional tests fails.

**Comment by pierreprinetti** [comment]
Can you please rebase? @mandre has been fixing most of the CI errors in the master branch.

**Comment by vooon** [comment]
@pierreprinetti done.

**Comment by pierreprinetti** [comment]
@vooon we may have issues here: internal/acceptance/openstack/containerinfra/v1/nodegroups_test.go

File: testhelper/convenience.go [hunk 1]
```diff
@@ -33,10 +33,12 @@ func yellow(str interface{}) string {
 }
 
 func logFatal(t *testing.T, str string) {
+	t.Helper()
```
Comment (mandre) [inline testhelper/convenience.go:36]: Nice, but this should be a separate PR that we backport.

**Review by mandre — COMMENTED** [review]
This looks great, thanks for doing this @vooon !

I'm sorry, I have merged https://github.com/gophercloud/gophercloud/pull/2994 prior to reviewing your PR. Could you once more rebase on master and update the newly introduced function to match your changes?

File: internal/acceptance/tools/tools.go [hunk 2]
```diff
@@ -1,39 +1,38 @@
...
-		satisfied, err := predicate()
-		if err != nil {
-			return err
-		}
-		if satisfied {
-			return nil
-		}
+	err := gophercloud.WaitFor(ctx, predicate)
+	if errors.Is(err, context.DeadlineExceeded) {
+		return ErrTimeout
```
[the remark this answers is not in the document]
Comment (vooon) [reply internal/acceptance/tools/tools.go:32]: I think i saw ErrTimeout usage somewhere, but didn't wanted to change acceptance more.
But now i actually do not see any occurrence.

File: internal/acceptance/tools/tools.go [hunk 3]
```diff
@@ -1,39 +1,38 @@
...
 	"time"
+
+	"github.com/gophercloud/gophercloud/v2"
 )
 
 // ErrTimeout is returned if WaitFor/WaitForTimeout take longer than their timeout duration.
 var ErrTimeout = errors.New("Timed out")
 
 // WaitFor uses WaitForTimeout to poll a predicate function once per second to
 // wait for a certain state to arrive, with a default timeout of 300 seconds.
```
[the remark this answers is not in the document]
Comment (vooon) [reply internal/acceptance/tools/tools.go]: I've changed help, but don't want to rename functions here.

File: testhelper/convenience.go (hunk 1 above)
Comment (vooon) [reply testhelper/convenience.go]: Done #2997

**Review by mandre — APPROVED** [review]
Thanks so much!

**Comment by pierreprinetti** [comment]
Nice! Thank you @vooon and @mandre for the additional review. I think we can merge as soon as we get the CI's blessing

**Review by pierreprinetti — APPROVED** [review]
